jobs/cloud_controller_ng/monit

```diff
@@ -2,18 +2,18 @@ check process cloud_controller_ng
   with pidfile /var/vcap/sys/run/bpm/cloud_controller_ng/cloud_controller_ng.pid
   start program "/var/vcap/jobs/bpm/bin/bpm start cloud_controller_ng"
   stop program "/var/vcap/jobs/bpm/bin/bpm stop cloud_controller_ng"
-  depends on ccng_monit_http_healthcheck
   group vcap
   if totalmem > <%= p("cc.thresholds.api.alert_if_above_mb") %> Mb for 3 cycles then alert
   if totalmem > <%= p("cc.thresholds.api.restart_if_consistently_above_mb") %> Mb for <%= p("cc.thresholds.api.restart_if_consistently_above_mb_cycles") %> cycles then exec "/var/vcap/jobs/cloud_controller_ng/bin/restart_drain"
   if totalmem > <%= p("cc.thresholds.api.restart_if_above_mb") %> Mb for 3 cycles then exec "/var/vcap/jobs/cloud_controller_ng/bin/restart_drain"
 
-
 check process ccng_monit_http_healthcheck
-    with pidfile /var/vcap/sys/run/bpm/cloud_controller_ng/ccng_monit_http_healthcheck.pid
-    start program "/var/vcap/jobs/bpm/bin/bpm start cloud_controller_ng -p ccng_monit_http_healthcheck"
-    stop program "/var/vcap/jobs/bpm/bin/bpm stop cloud_controller_ng -p ccng_monit_http_healthcheck"
-    group vcap
+  with pidfile /var/vcap/sys/run/bpm/cloud_controller_ng/ccng_monit_http_healthcheck.pid
+  start program "/var/vcap/jobs/bpm/bin/bpm start cloud_controller_ng -p ccng_monit_http_healthcheck"
+  stop program "/var/vcap/jobs/bpm/bin/bpm stop cloud_controller_ng -p ccng_monit_http_healthcheck"
+  if 1 restart within 2 cycles then exec "/var/vcap/jobs/cloud_controller_ng/bin/restart_drain"
+  depends on nginx_cc
```

Comment:
The `restart_drain` script does two things to prevent any interference with `monit`:

1. It calls `monit unmonitor` [here](https://github.com/cloudfoundry/capi-release/blob/7451e04af4689d655720fcab44d5271600c1869c/jobs/cloud_controller_ng/templates/restart_drain.sh.erb#L22-L25). This prevents `monit` from taking any actions in the following cycles. The same is done when BOSH stops a VM; it first calls `monit unmonitor` for all jobs and then the `drain` script of the jobs (if provided).

2. It uses a PID guard to ensure that there is no a parallel execution of this script ([here](https://github.com/cloudfoundry/capi-release/blob/7451e04af4689d655720fcab44d5271600c1869c/jobs/cloud_controller_ng/templates/restart_drain.sh.erb#L22-L25)).

So the scenario you described (totalmem > threshhold) will result in the following:
1. `monit` calls the `restart_drain` script,
2. `restart_drain` calls `monit unmonitor`,
3. graceful shutdown,
4. `restart_drain` calls `monit monitor`,
5. `monit` restarts the processes (as they are stopped)

There still might be an edge case that could lead to `restart_drain` being called twice. The `totalmem` (cloud_controller_ng) and the `restart` (ccng_monit_http_healthcheck) checks could be triggered independently at the same time (i.e. both conditions are true in the same `monit` cycle). Then the PID guard mentioned above should ensure that one invocation triggers the graceful shutdown and the other one just returns.

Reply:
Got it.  Thanks for the thorough explanation.